vpr/SRC/route/route_timing.c

```diff
@@ -514,27 +556,114 @@ bool timing_driven_route_net(int inet, int itry, float pres_fac, float max_criti
 	 * case the rr_graph is disconnected and you can give up. If slacks = NULL, *
 	 * give each net a dummy criticality of 0.									*/
 
-	unsigned int ipin;
-	int num_sinks, itarget, target_pin, target_node, inode;
-	float target_criticality, old_total_cost, new_total_cost, largest_criticality,
-		old_back_cost, new_back_cost;
-	t_rt_node *rt_root;
-	
-	struct s_trace *new_route_start_tptr;
-	int highfanout_rlim;
+	// for nets below a certain size (min_incremental_reroute_fanout), rip up any old routing
+	// otherwise, we incrementally reroute by reusing legal parts of the previous iteration
+	// convert the previous iteration's traceback into the starting route tree for this iteration
 
-	/* Rip-up any old routing. */
+	INCREMENTAL_REROUTING_ASSERT(sanity_check_lookup());
 
-	pathfinder_update_one_cost(trace_head[inet], -1, pres_fac);
-	free_traceback(inet);
-	
-	for (ipin = 1; ipin < g_clbs_nlist.net[inet].pins.size(); ipin++) {
-		if (!slacks) {
-			/* Use criticality of 1. This makes all nets critical.  Note: There is a big difference between setting pin criticality to 0
-			compared to 1.  If pin criticality is set to 0, then the current path delay is completely ignored during routing.  By setting
-			pin criticality to 1, the current path delay to the pin will always be considered and optimized for */
-			pin_criticality[ipin] = 1.0;
-		} else { 
+	t_rt_node* rt_root;
+	remaining_targets.clear();	// efficient clearing by just resetting size without allocation
+	unsigned int num_sinks = g_clbs_nlist.net[inet].num_sinks();
+
+	if ((int)num_sinks < min_incremental_reroute_fanout || itry == 1) {
+#ifdef PROFILE
+		++entire_net_rerouted;
+#endif
+		// rip up the whole net
+		pathfinder_update_one_cost(trace_head[inet], -1, pres_fac);
+		free_traceback(inet);
+
+		rt_root = init_route_tree_to_source(inet);
+		for (unsigned int sink_pin = 1; sink_pin <= num_sinks; ++sink_pin)
+			remaining_targets.push_back(sink_pin);
+		// when we don't prune the tree, we also don't know the sink node indices 
+		// thus we'll use functions that act on pin indices like mark_ends instead
+		// of their versions that act on node indices directly like mark_remaining_ends
+		mark_ends(inet);
+	}
+	else {
+		reached_sinks.clear();
+#ifdef PROFILE
+		clock_t rebuild_start = clock();
+#endif
+		// convert the previous iteration's traceback into the partial route tree for this iteration
+		rt_root = traceback_to_route_tree(inet);
+
+		// check for edge correctness
+		INCREMENTAL_REROUTING_ASSERT(is_valid_skeleton_tree(rt_root));
+		INCREMENTAL_REROUTING_ASSERT(should_route_net(inet));
+		INCREMENTAL_REROUTING_ASSERT(!is_uncongested_route_tree(rt_root));
+
+		// prune the branches of the tree that don't legally lead to sinks
+		// destroyed represents whether the entire tree is illegal
+		bool destroyed {prune_route_tree(rt_root, pres_fac, reached_sinks, remaining_targets)};

```

Comment:
{} assignment warns against conversions that = allows (silently). I think it's generally a Good Thing, but you're right here that bool assigned to a function returning bool shouldn't have conversion problems.